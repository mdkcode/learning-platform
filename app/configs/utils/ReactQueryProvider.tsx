"use client";
import { useState } from "react";
import {
  QueryClientProvider,
  QueryClient,
  QueryCache,
} from "@tanstack/react-query";
import { useErrorHandler } from "@/app/configs/hooks/useErrorHandler";

interface ReactQueryProviderProps {
  children?: React.ReactNode;
}

export const ReactQueryProvider: React.FC<ReactQueryProviderProps> = ({
  children,
}) => {
  const onError = useErrorHandler();
  const [queryClient] = useState(
    () =>
      new QueryClient({
        queryCache: new QueryCache({
          onError,
        }),
      })
  );

  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
};
