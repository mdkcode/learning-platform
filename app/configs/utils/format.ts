import { Timestamp } from "firebase/firestore";

export function convertTimestampToDateString(created_at: Timestamp) {
  return created_at ? created_at.toDate().toISOString() : "";
}

export function convertSecondsToMinutes(seconds: number) {
  const minutes = Math.ceil(seconds / 60);
  return minutes;
}
