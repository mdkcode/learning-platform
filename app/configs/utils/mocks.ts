export const mockUser = {
  name: "Jane Doe",
  email: "jane.doe@example.com",
  image: "https://example.com/avatar.jpg",
};

export const mockUserSession = {
  user: mockUser,
  expires: new Date().toISOString(),
  accessToken: "dsbhkjscd",
  refreshToken: "894655sdc",
  userId: "1111",
};
