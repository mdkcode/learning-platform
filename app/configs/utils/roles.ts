export enum Roles {
  Student = "STUDENT",
  Admin = "ADMIN",
}
