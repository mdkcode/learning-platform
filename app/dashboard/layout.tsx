"use client";
import { useSession } from "next-auth/react";
import { MobileHeader } from "../ui/dashboard/MobileHeader";
import SideNav from "../ui/dashboard/sidenav";
import { useRouter } from "next/navigation";
import { useEffect } from "react";
import { Routes } from "../routes/routes";

const Layout = ({ children }: { children: React.ReactNode }) => {
  const router = useRouter();
  const { data: session } = useSession();

  useEffect(() => {
    if (session?.expires && !session.user) router.push(Routes.HOME);
  }, [session?.expires, session?.user, router]);

  return (
    <div className="flex flex-col md:h-screen md:flex-row">
      <MobileHeader />
      <SideNav />
      <div className="flex-grow p-6 md:overflow-y-auto md:p-12">{children}</div>
    </div>
  );
};

export default Layout;
