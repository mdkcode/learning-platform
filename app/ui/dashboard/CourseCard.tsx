"use client";
import { getCourseList } from "@/app/api/courses/courses.api";
import {
  CourseProps,
  CourseSearchParams,
} from "@/app/api/courses/courses.interface";
import {
  DehydratedState,
  HydrationBoundary,
  useQuery,
} from "@tanstack/react-query";

interface CourseCardsListClientProps extends CourseSearchParams {
  dehydratedState: DehydratedState;
}

export const CourseCardsList = ({
  searchQuery,
  dehydratedState,
}: CourseCardsListClientProps) => {
  const { data: videos } = useQuery({
    queryKey: ["courses", searchQuery],
    queryFn: () => getCourseList(searchQuery),
  });

  return (
    <HydrationBoundary state={dehydratedState}>
      <div className="flex gap-5 flex-wrap">
        {videos?.map((video: CourseProps) => (
          <CourseCard key={video.id} {...video} />
        ))}
      </div>
    </HydrationBoundary>
  );
};

export const CourseCard = ({ name, description }: CourseProps) => {
  return (
    <div
      className="w-[400px] inline-block rounded-lg p-5 bg-gradient-to-b bg-gray-200 
      relative cursor-pointer transition-all duration-300 transform 
      hover:scale-105 hover:shadow-lg hover:from-gray-200"
    >
      <h3 className="text-xl font-bold my-2">{name}</h3>
      <p className="mb-8">{description}</p>
      <div className="absolute bottom-3">
        <p className="text-lg text-blue-500">Subscribe</p>
      </div>
    </div>
  );
};
