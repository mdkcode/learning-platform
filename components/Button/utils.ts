export enum ButtonType {
  PRIMARY = "primary",
  SECONDARY = "secondary",
  NAVIGATION = "navigation",
}

export const defaultClassName = {
  [ButtonType.PRIMARY]:
    "px-7 py-3 bg-[#00004B] text-white font-semibold rounded-lg shadow-md hover:bg-[#3A59FF] focus:outline-none focus:ring-2 focus:ring-[#3A59FF]",
  [ButtonType.SECONDARY]:
    "px-7 py-3 border-2 border-solid border-[#00004B] bg-white text-[#00004B] font-semibold rounded-lg hover:text-[#3A59FF] hover:border-[#3A59FF]",
  [ButtonType.NAVIGATION]:
    "px-7 py-3 text-white font-semibold rounded-lg bg-gradient-to-r from-orange-400 to-pink-500 hover:bg-gradient-to-r hover:from-orange-500 hover:to-pink-600",
};
