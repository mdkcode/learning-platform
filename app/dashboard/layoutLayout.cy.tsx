import React from "react";
import Layout from "./layout";
import "@/app/globals.css";
import { AppRouterContext } from "next/dist/shared/lib/app-router-context.shared-runtime";
import { SessionProvider } from "next-auth/react";
import * as nextRouter from "next/navigation";
import { mockUserSession } from "@/app/configs/utils/mocks";

describe("<Layout />", () => {
  it("should redirect to home if the session is invalid", () => {
    const router = {
      push: cy.stub().as("router:push"),
      back: cy.stub(),
      forward: cy.stub(),
      refresh: cy.stub(),
      replace: cy.stub(),
      prefetch: cy.stub(),
    };
    // Mock useRouter to return the mocked router
    cy.stub(nextRouter, "useRouter").returns(router);

    cy.mount(
      <AppRouterContext value={router}>
        <SessionProvider session={mockUserSession}>
          <Layout>
            <div>Test Child</div>
          </Layout>
        </SessionProvider>
      </AppRouterContext>
    );
    cy.contains("Test Child").should("exist");
    cy.get('[data-test="mobileHeader"]').should("exist");
    cy.get('[data-test="sidenav"]').should("exist");
  });
});
