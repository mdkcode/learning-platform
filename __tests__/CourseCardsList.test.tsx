import { render } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import { CourseCardsList, CourseCard } from "@/app/ui/dashboard/CourseCard";
import { useQuery } from "@tanstack/react-query";
import { SessionProvider } from "next-auth/react";
import { mockUserSession } from "@/app/configs/utils/mocks";

vi.mock("@tanstack/react-query", () => ({
  useQuery: vi.fn(),
  useMutation: vi.fn(() => ({
    mutate: vi.fn(),
  })),
}));

vi.mock("react-toastify", () => ({
  toast: { success: vi.fn() },
}));

describe("CourseCardsList", () => {
  it("matches snapshot with courses", () => {
    (useQuery as ReturnType<typeof vi.fn>).mockReturnValue({
      data: [
        {
          id: "course-1",
          name: "React Basics",
          description: "Learn the basics of React.js",
        },
        {
          id: "course-2",
          name: "Advanced TypeScript",
          description: "Deep dive into TypeScript features",
        },
      ],
    });

    const { container } = render(
      <SessionProvider session={mockUserSession}>
        <CourseCardsList searchQuery="React" />
      </SessionProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with no courses", () => {
    (useQuery as ReturnType<typeof vi.fn>).mockReturnValue({
      data: [],
    });

    const { container } = render(
      <SessionProvider session={mockUserSession}>
        <CourseCardsList searchQuery="React" />
      </SessionProvider>
    );
    expect(container).toMatchSnapshot();
  });
});

describe("CourseCard", () => {
  it("matches snapshot with a course", () => {
    const { container } = render(
      <SessionProvider session={mockUserSession}>
        <CourseCard
          id="course-1"
          name="React Basics"
          description="Learn the basics of React.js"
          image_url="/"
          created_at=""
          duration={5}
        />
      </SessionProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
