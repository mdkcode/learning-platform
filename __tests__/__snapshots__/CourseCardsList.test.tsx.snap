// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CourseCard > matches snapshot with a course 1`] = `
<div>
  <div
    class="w-[400px] inline-block rounded-lg p-5 bg-gradient-to-b bg-gray-200 
      relative cursor-pointer transition-all duration-300 transform 
      hover:scale-105 hover:shadow-lg hover:from-gray-200"
  >
    <h3
      class="text-xl font-bold my-2"
    >
      React Basics
    </h3>
    <p
      class="mb-8"
    >
      Learn the basics of React.js
    </p>
    <div
      class="absolute bottom-3"
    >
      <p
        class="text-lg text-blue-500"
      >
        Subscribe
      </p>
    </div>
  </div>
</div>
`;

exports[`CourseCardsList > matches snapshot with courses 1`] = `
<div>
  <div
    class="flex gap-5 flex-wrap"
  >
    <div
      class="w-[400px] inline-block rounded-lg p-5 bg-gradient-to-b bg-gray-200 
      relative cursor-pointer transition-all duration-300 transform 
      hover:scale-105 hover:shadow-lg hover:from-gray-200"
    >
      <h3
        class="text-xl font-bold my-2"
      >
        React Basics
      </h3>
      <p
        class="mb-8"
      >
        Learn the basics of React.js
      </p>
      <div
        class="absolute bottom-3"
      >
        <p
          class="text-lg text-blue-500"
        >
          Subscribe
        </p>
      </div>
    </div>
    <div
      class="w-[400px] inline-block rounded-lg p-5 bg-gradient-to-b bg-gray-200 
      relative cursor-pointer transition-all duration-300 transform 
      hover:scale-105 hover:shadow-lg hover:from-gray-200"
    >
      <h3
        class="text-xl font-bold my-2"
      >
        Advanced TypeScript
      </h3>
      <p
        class="mb-8"
      >
        Deep dive into TypeScript features
      </p>
      <div
        class="absolute bottom-3"
      >
        <p
          class="text-lg text-blue-500"
        >
          Subscribe
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`CourseCardsList > matches snapshot with no courses 1`] = `
<div>
  <div
    class="flex gap-5 flex-wrap"
  />
</div>
`;
